### ISSUE TITLE
Browser: Same Origin Policy

### ISSUE THREAD

**Issue by bcoles** [SerenityOS/serenity#1615]
The `Browser` application and associated `LibJS` do not yet enforce the Same Origin Policy. It is probably time to start thinking about it.

**Comment by bcoles** [comment]
Support for document origin `Document::origin()` has been implemented in 7382b27c220327470a84a2e920277be93224a6d3, but is not yet enforced.

**Comment by bcoles** [comment]
A rudimentary patch:

```diff
diff --git a/Libraries/LibWeb/DOM/XMLHttpRequest.cpp b/Libraries/LibWeb/DOM/XMLHttpRequest.cpp
index 82caf5b71..2dfe049ea 100644
--- a/Libraries/LibWeb/DOM/XMLHttpRequest.cpp
+++ b/Libraries/LibWeb/DOM/XMLHttpRequest.cpp
@@ -61,8 +61,14 @@ void XMLHttpRequest::open(const String& method, const String& url)
 
 void XMLHttpRequest::send()
 {
+    auto origin_url = m_window->document().url();
+    auto request_url = m_window->document().complete_url(m_url);
+    if (origin_url.protocol() != request_url.protocol() || origin_url.host() != request_url.host() || origin_url.port() != request_url.port()) {
+        dbg() << "XHR failed to load: forbidden to load " << request_url.to_string() << " from " << origin_url.to_string();
+        return;
+    }
     ResourceLoader::the().load(
-        m_window->document().complete_url(m_url),
+        request_url,
         [weak_this = make_weak_ptr()](auto& data) {
             if (!weak_this)
                 return;
```

This blocks cross-origin requests outright, which doesn't align with other browsers. Other browsers permit the request, but deny access to response object properties (such as `data`) if the request was cross-origin.
